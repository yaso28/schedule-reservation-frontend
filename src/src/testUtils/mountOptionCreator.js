import { createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import routes from '@/router/routes'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import i18nOptions from '@/i18n/options'
import VueMeta from 'vue-meta'
import CoreuiVue from '@coreui/vue'
import icons from '@/icons'
import services from '@/services'

class MountOptionCreator {
  create (storeModules = null) {
    const localVue = createLocalVue()

    // router
    localVue.use(VueRouter)
    const router = new VueRouter({
      mode: 'history',
      base: process.env.BASE_URL,
      routes
    })

    // i18n
    localVue.use(VueI18n)
    const i18n = new VueI18n(i18nOptions)

    // vue-meta
    localVue.use(VueMeta)

    // Core UI
    localVue.use(CoreuiVue)

    const option = {
      localVue,
      router,
      i18n,
      icons,
      provide: services
    }

    // store
    if (storeModules) {
      localVue.use(Vuex)
      option.store = new Vuex.Store({
        modules: storeModules
      })
    }

    return option
  }
}

export default new MountOptionCreator()
