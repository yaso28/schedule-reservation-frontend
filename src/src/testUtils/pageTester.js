import { mount } from '@vue/test-utils'
import MyPage from '@/components/containers/MyPage'

class PageTester {
  async mount (pageComponent, route, mountOption) {
    await mountOption.router.push(route)
    const pageWrapper = mount(MyPage, mountOption)
    return pageWrapper.findComponent(pageComponent)
  }

  assertTitleTag () {
    expect(document.title).toMatchSnapshot()
  }

  assertMetaTags () {
    expect(document.querySelectorAll('meta[data-vue-meta]')).toMatchSnapshot()
  }
}

export default new PageTester()
