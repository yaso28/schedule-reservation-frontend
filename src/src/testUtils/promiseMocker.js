import _ from 'lodash'

const createSetTimeoutPromise = getResult => new Promise((resolve, reject) => {
  setTimeout(() => getResult(resolve, reject), 1000)
})
const createSwitchPromise = (value, error, switchObj) => createSetTimeoutPromise(
  (resolve, reject) => switchObj.isSuccess ? resolve(value) : reject(error)
)
const createResolvedPromise = value => createSetTimeoutPromise(resolve => resolve(value))
const createRejectedPromise = error => createSetTimeoutPromise((resolve, reject) => reject(error))

const createMock = createPromise => jest.fn().mockImplementationOnce(createPromise)
const createMockForCases = (caseList, createPromiseForCurrentCase) => jest.fn((...args) => {
  for (let i = 0; i < caseList.length; i++) {
    const currentCase = caseList[i]
    if (_.isEqual(args, currentCase.args)) {
      return createPromiseForCurrentCase(currentCase)
    }
  }
  throw new Error('createMockForCases: no args matched')
})

class PromiseMocker {
  init () {
    jest.useFakeTimers()
  }

  clear () {
    jest.clearAllTimers()
  }

  flushPendings () {
    jest.runOnlyPendingTimers()
  }

  createSwitch () {
    return { isSuccess: true }
  }

  setSwitch (switchObj, isSuccess) {
    switchObj.isSuccess = isSuccess
  }

  mock (value, error, switchObj) {
    return createMock(() => createSwitchPromise(value, error, switchObj))
  }

  mockResolved (value) {
    return createMock(() => createResolvedPromise(value))
  }

  mockRejected (error) {
    return createMock(() => createRejectedPromise(error))
  }

  mockForCases (caseList) {
    return createMockForCases(caseList, currentCase => createSwitchPromise(currentCase.value, currentCase.error, currentCase.switch))
  }

  mockResolvedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createResolvedPromise(currentCase.value))
  }

  mockRejectedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createRejectedPromise(currentCase.error))
  }
}

export default new PromiseMocker()
