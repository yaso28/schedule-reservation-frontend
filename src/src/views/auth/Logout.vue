<template>
  <div />
</template>

<script>
export default {
  name: 'AuthLogout',
  inject: ['apiService'],
  inheritAttrs: false,
  mounted () {
    this.logout()
  },
  methods: {
    async logout () {
      try {
        this.$store.commit('processing/beginPost')
        await this.apiService.logout()
        this.$store.dispatch('auth/clear')
        this.$router.push({ name: 'login' })
      } catch (e) {
        if (!this.apiService.isHandledError(e)) {
          throw e
        }
      } finally {
        this.$store.commit('processing/endPost')
      }
    }
  }
}
</script>
