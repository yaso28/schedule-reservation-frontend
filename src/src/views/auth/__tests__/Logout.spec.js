import Logout from '../Logout'
import mountOptionCreator from '@/testUtils/mountOptionCreator'
import pageTester from '@/testUtils/pageTester'
import promiseMocker from '@/testUtils/promiseMocker'
import storeModules from '@/store/modules'

jest.mock('@/store/modules')

describe('Logout', () => {
  let mountOption
  let router
  let apiService
  beforeEach(() => {
    promiseMocker.init()
    mountOption = mountOptionCreator.create(storeModules)
    router = mountOption.router
    apiService = mountOption.provide.apiService
  })

  afterEach(() => {
    promiseMocker.clear()
    jest.resetAllMocks()
  })

  describe('マウント', () => {
    let wrapper
    let logoutSwitch
    let logoutError
    beforeEach(async () => {
      logoutSwitch = promiseMocker.createSwitch()
      logoutError = new Error()
      apiService.logout = promiseMocker.mock(
        null,
        logoutError,
        logoutSwitch
      )
      wrapper = await pageTester.mount(Logout, { name: 'logout' }, mountOption)
    })

    test('titleタグ', () => {
      pageTester.assertTitleTag()
    })

    test('metaタグ', () => {
      pageTester.assertMetaTags()
    })

    test('ページ全体表示', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    test('Processingカバー表示開始', () => {
      expect(storeModules.processing.mutations.beginPost).toHaveBeenCalled()
    })

    test('ログアウト処理を呼出', () => {
      expect(apiService.logout).toHaveBeenCalledWith()
    })

    test('Processingカバー表示中', () => {
      expect(storeModules.processing.mutations.endPost).not.toHaveBeenCalled()
    })

    describe('ログアウト：エラー', () => {
      beforeEach(() => {
        promiseMocker.setSwitch(logoutSwitch, false)
        apiService.isHandledError = jest.fn().mockReturnValueOnce(true)
        promiseMocker.flushPendings()
      })

      test('エラー処理', () => {
        expect(apiService.isHandledError).toHaveBeenCalledWith(logoutError)
      })

      test('Processingカバー表示終了', () => {
        expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
      })
    })

    describe('ログアウト：成功', () => {
      beforeEach(() => {
        promiseMocker.setSwitch(logoutSwitch, true)
        router.push = jest.fn()
        promiseMocker.flushPendings()
      })

      test('認証情報クリア', () => {
        expect(storeModules.auth.actions.clear).toHaveBeenCalled()
      })

      test('ログインページへ転送', () => {
        expect(router.push).toHaveBeenCalledWith({ name: 'login' })
      })

      test('Processingカバー表示終了', () => {
        expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
      })
    })
  })
})
