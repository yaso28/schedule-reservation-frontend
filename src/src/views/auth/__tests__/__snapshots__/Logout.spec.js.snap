// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout マウント metaタグ 1`] = `
NodeList [
  <meta
    content="noindex"
    data-vmid="robots"
    data-vue-meta="2"
    name="robots"
  />,
]
`;

exports[`Logout マウント titleタグ 1`] = `"やそスポーツクラブ"`;

exports[`Logout マウント ページ全体表示 1`] = `<div />`;
